src/visualizers/VisualizerGrid.ts

```diff
@@ -0,0 +1,959 @@
+import {VisualizerExportModule} from '@/visualizers/VisualizerInterface'
```

Comment:
(I'm commenting on a file so we can hide the convo if it becomes resolved and PR won't look so cluttered.)

Glen and I had a discussion in the previous PR about performance.

Glen: Is it just me, or is this visualizer really slow? I used the natural numbers, and just one property "divisible by seven", and didn't touch anything else and the picture took quite a number of seconds to show up -- a really noticeable delay, which is bad for the user experience. But it seems to me that testing 40000 numbers for divisibility by 7 should not take anywhere near that long... How performant does it seem for the rest of you with those settings?

Liam: It is also slow for me. One simple way to test wall clock time would be to put `console.time` and `console.timeEnd` statements in the code and see what the bottleneck(s) is(are).

Glen: Right well once we work through the outstanding items, we can also just take a look for any obvious inefficiencies; and of course, we can also make the Visualizer not draw the entire canvas at once, but instead do a certain number of points per frame. The latter won't speed anything up, but it will vastly improve responsiveness, which is arguably more important than raw speed.

Glen: Let's wait on changing the default number of terms to 4096 until someone has a chance to look at the speed... if it turns out to be easy to speed up, we can leave the default as it is, since as far as I can tell the only motivation for changing the default was the performance.

Reply:
Oops, I didn't notice this when I was looking for a quick change to test my setup.  Now it's at 4096, so someone will have to change it back if we speed it up.

Reply:
Hi all, so I took a look at this.  I tried to move things from the iteration into the setup.  I think this helped a bit (e.g. no longer drawing background squares each time, but just setting a background first).  I put a console message at the beginning of the iteration, and observed something very weird.  I used Glen's suggestion of divisible by 7 for 40000 numbers.  The message appeared immediately (meaning the `setup()` was fast) and then there was a long delay.  Then the message appeared _again_ and it drew almost immediately.  What is going on?  It appears the thumbnail card is calling `draw()` twice (there's a `noLoop()` in `draw()`), and the first time is taking the long time, but the second time is fast.  Going to the big canvas is fast too

Reply:
Ok, update:  the bottleneck is `seq.getElement()` in Glen's example.  When I made a dummy to replace that (just making all terms `=5n`), it was 10x faster.  So that explains why the first time through `draw()` is long and subsequent times are fast.  I tried using sequence A000005, which has 100000 terms in the OEIS, to see if `seq.getElement()` is faster there.  But I can't seem to get the backend to cache it all.  I'll check back later.

Reply:
Ok, I was finally able to verify with a023900, which has 20000 terms, that the visualizer itself is twice as fast on an OEIS sequence than on a sequence by formula; that's what's slow.  I think this isn't an issue with this PR, so I suggest we discuss and file an issue, then resolve this convo.

Reply:
I don't understand what the issue is, so I'd like to discuss this on Monday. We can write up an issue and resolve this thread if it's not an issue with this PR.

Reply:
Maybe somehow caching and recompiling is slow on sequences by formula? Like there's some accidental quadratic term in sequences by formula? I don't know that's true. Just trying to brainstorm a counterintuitive result.